### ISSUE TITLE
Regression: Module parameters lost

### ISSUE THREAD

**Issue by WolframKahl** [agda/agda#1701]
With the development version 759a4dc500629270ec04c378a724a038217a439f, a module parameter passing pattern that I used before successfully does not work anymore. I managed to produce a trimmed-down example showing this effect (with the hole filled, it typechecks using Agda-2.4.2.2):

``` agda
module ModParamsLost where

  module ModParamsRecord (A : Set) where
    record R (B : Set) : Set where
      field F : A → B

  module ModParamsToLoose (A : Set) where
    open ModParamsRecord

    module _ (B : Set) (G : A → B) where
      r : R A B
      r = record { F = G }

      module r = R A r

  module ModParamsLost (A : Set) where
    open ModParamsRecord
    open ModParamsToLoose A

    f : (A → A) → A → A
    f G = {! S.F !}
      where
        module S = r A G           -- expected |S.F : A → A|,
                                   -- but obtained |S.F : (B : Set) (G₁ : A → B) → A → B|
        -- module S = r            -- as expected: |S.F : (B : Set) (G₁ : A → B) → A → B|
        -- module S = R A (r A G)  -- as expected: |S.F : A → A|
```

The typing appears to pretend that the parameters `A` and `G` are not there in `module S = r A G`, which is what I want to write, and what worked before.

**Comment by andreasabel** [comment]
`git bisect` thinks the bug was introduced by commit cd0c777b0594a12da5954f5247693a1d759818c4 fixing #1396

```
commit cd0c777b0594a12da5954f5247693a1d759818c4
Author: Ulf Norell <ulf.norell@gmail.com>
Date:   Wed Mar 11 16:21:28 2015 +0100

    [ issue 1396 ] fixed bug when applying modules with parameterised submodules
```

**Comment by andreasabel** [comment]
The problem is that the free variable computation in `TypeChecking.Monad.Signature`

``` haskell
    argsToUse new = do
      let m = mnameFromList $ commonPrefix (mnameToList old) (mnameToList new)
      reportSLn "tc.mod.apply" 80 $ "Common prefix: " ++ show m
      let ms = map mnameFromList . tail . inits . mnameToList $ m  -- NB: tail . inits computes the non-empty prefixes
      ps <- mapM (fmap secFreeVars <.> getSection) ms
      reportSLn "tc.mod.apply" 80 $ "  params: " ++ show (zip ms ps)
      return $ sum $ map (fromMaybe 0) ps
```

does not work properly if `getSection` returns `Nothing` because the module has not been copied.  This holds for private and (so far) anonymous modules.  Here is a variation of the test case:

``` agda
module _ where

  module ModParamsRecord (A : Set) where
    record R (B : Set) : Set where
      field F : A → B

  module ModParamsToLoose (A : Set) where
    open ModParamsRecord

    private
      module M (B : Set) (G : A → B) where
        r : R A B
        r = record { F = G }

        module r = R A r
    open M public

  module ModParamsLost (A : Set) where
    open ModParamsRecord
    open ModParamsToLoose A

    f : (A → A) → A → A
    f G = S.F
      where
        module S = r A G      
```

This works if `private` is removed.

**Comment by andreasabel** [comment]
@vlopezj : Another interesting bug concerning the module system.

**Comment by andreasabel** [comment]
Here is a further shrunk test case without records:

``` agda
postulate
  A : Set
  a : A

module Id (A : Set) (a : A) where
  x = a

module ModParamsToLoose (A : Set) where
  private module M (a : A) where module I = Id A a
  open M public

open ModParamsToLoose A
open I a

y : A
y = x
```

Agda thinks `x : (a1 : A) -> A`.

**Comment by andreasabel** [comment]
81e99cf triggers in http://relmics.mcmaster.ca/RATH-Agda/AContext-2.1.tar.gz

```
Checking Categoric.SGFunctor.Coproduct (/var/tmp/kahl/svn/C/RATH-Agda/trunk/Categoric/SGFunctor/Coproduct.lagda).
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Monad/Signature.hs:563
```

**Comment by andreasabel** [comment]
Shrunk test case:

``` agda
postulate N : Set

module A (O : Set) where

  record R : Set where
    field f : O

  module L where
    open R public

module B (P : Set) where

  module M (Q : Set) where
    open A Q public

  open M P public

open B
open L N
```

**Comment by WolframKahl** [comment]
On Tue, Nov 24, 2015 at 10:49:04AM -0800, Andreas Abel wrote:

> Closed #1701.

I found a fix in master (and it works for Categoric/SGFunctor/Coproducts.lagda,
but apparently not for the problem behind the original error report for this issue)
but not yet in  https://github.com/andreasabel/agda-issue-1692.git ---
which way will this move? 2.4.2.5 from agda-issue-1692.git
appears to have a problem with at least one of my `with`s.

Wolfram

**Comment by WolframKahl** [comment]
With master at 36a35e3858d4379e92ddd5d3db835e0ff766938c , I get a new module system regression:

```
/var/tmp/kahl/svn/C/RATH-Agda/trunk/Data/SUList/SepElemTypes.lagda:1484,22-50
Core.SUList Key Elem key m₁ !=<
M2a′.Core₁.SUList Elem key Elem key m₁ of type
Set (ℓE ⊍ (ℓk₂ ⊍ (ℓk₁ ⊍ ℓK)))
when checking that the expression M2a′.IW-reflexive′ Q QK R RK has
type
({e₁ e₂ : Elem} → R e₁ e₂ → Q e₂ e₁) →
{m₁ m₂ : Maybe K} {es₁ : SUList m₁} {es₂ : SUList m₂} →
EquivWith es₁ es₂ → InclWith es₂ es₁
```

I put the development up to that file at:
http://relmics.mcmaster.ca/RATH-Agda/SUList-Core-2015-11-24.tar.gz

**Comment by WolframKahl** [comment]
With master still at 36a35e3858d4379e92ddd5d3db835e0ff766938c, the original `ModParamsLost` starting this issue is indeed accepted, but the file that motivated this issue is not:

```
/var/tmp/kahl/svn/C/RATH-Agda/trunk/Categoric/MonoidalCategory/Outer/Wire.lagda:122,70-76
((Φ₁
  : .Categoric.MonoidalCategory.Outer.Comp._.Mor F ➦pushout
    ➦pushout-presMono (_A_249 F ➦pushout ➦pushout-presMono)
    (_B_250 F ➦pushout ➦pushout-presMono))
 (Ψ
  : .Categoric.MonoidalCategory.Outer.Comp._.Mor F ➦pushout
    ➦pushout-presMono (_B_250 F ➦pushout ➦pushout-presMono)
    (_C_251 F ➦pushout ➦pushout-presMono)) →
 .Categoric.FinColimits.Pushout.IsPushout compOp₂
 (.Categoric.MonoidalCategory.Outer.Comp._.Φ.trgConn F ➦pushout
  ➦pushout-presMono Φ₁ Ψ)
 (.Categoric.MonoidalCategory.Outer.Base.Mor.srcConn Ψ)
 (.Categoric.FinColimits.Pushout.Pushout.left (⨾-PO Φ₁ Ψ))
 (.Categoric.FinColimits.Pushout.Pushout.right (⨾-PO Φ₁ Ψ)))
!=<
(.Categoric.FinColimits.Pushout.IsPushout compOp₂
 (_F_246 F ➦pushout ➦pushout-presMono) 𝒞₂.Id
 (_R_247 F ➦pushout ➦pushout-presMono)
 (_S_248 F ➦pushout ➦pushout-presMono))
of type (Set (ℓk₂ ⊍ (ℓj₂ ⊍ ℓo₂)))
when checking that the expression PO.prf has type
.Categoric.Category.FinColimits.CatFinColimits₀.FinColimits.IsPushout
𝒞₂ (_F_246 F ➦pushout ➦pushout-presMono) 𝒞₂.Id
(_R_247 F ➦pushout ➦pushout-presMono)
(_S_248 F ➦pushout ➦pushout-presMono)
```

As I won't have time these days to attempt another slimming down (getting the original ModParamsLost was already pretty hard), I just put everything up to that file at: 
http://relmics.mcmaster.ca/RATH-Agda/MonoidalCategory-Outer-Wire-2015-11-24.tar.gz .
This includes the problematic file `Categoric/MonoidalCategory/Outer/Wire.lagda`,
which worked before,
and a simple work-around in `Categoric/MonoidalCategory/Outer/Wire2.lagda`,
which is currently accepted.

**Comment by andreasabel** [comment]
Regarding SepElemTypes, this does not check for me as Data.Maybe.\* is missing in your .tgz.

**Comment by andreasabel** [comment]
Partially shrank (in 4 hours) to http://www.cse.chalmers.se/~abela/Outer-Wire2.tgz, problematic file is still Categoric/MonoidalCategory/Outer/Wire.lagda.
Works with agda-2.4.2.2 and fails with agda-2.4.2.3.

**Comment by andreasabel** [comment]
Actually the original test case is as good if split over several files:

``` agda
module Issue1701.ModParamsRecord (A : Set) where

record R (B : Set) : Set where
  field F : A → B
```

``` agda
module Issue1701.ModParamsToLoose (A : Set) where

open import Issue1701.ModParamsRecord

module _ (B : Set) (G : A → B) where
  r : R A B
  r = record { F = G }

  module r = R A r
```

Main file:

``` agda
module _ (A : Set) where
open import Issue1701.ModParamsRecord
open import Issue1701.ModParamsToLoose A

f : (A → A) → A → A
f G = S.F
  where
    module S = r A G 
```

Originally, I fixed this issue by keeping private modules around.  However, through serialization, they are deleted anyway.
Trying to keep them in the serialized files, by modifying the `publicModules` function to not `restrictPrivate`, gives me clashes and panics about ambiguous names.  Currently, I am stuck.

**Comment by UlfNorell** [comment]
The problem arises when a parameterised parent module is private, and thus not getting copied. It isn't a scope issue though--the parent module really should be private--so instead of messing about with the scope I fixed this by always copying parent modules.

@andreasabel: I think you can rollback your scope changes (668c7ff2 and 80625047), but I have not checked.

**Comment by WolframKahl** [comment]
At b289cbf2ce34d0daf295ef2d2f71fe47ccbacfd6, the Wire problem is gone --- thank you!
But the SepElemTypes problem is still there...

**Comment by asr** [comment]
> @andreasabel: I think you can rollback your scope changes (668c7ff and 8062504), but I have not checked.

I reverted 8062504 by d5fd6f466a0ff8d543126c9015d5a860960d727c.

**Comment by andreasabel** [comment]
https://github.com/agda/agda/commit/668c7ff22ca102f8af8136c7ac11dcd456c5564b has been reverted by https://github.com/agda/agda/commit/80625047ebbf0fbf8f338ef876712f84ed076853 but unreverted by https://github.com/agda/agda/commit/d5fd6f466a0ff8d543126c9015d5a860960d727c and finally ununreverted by 7a0838cedaeb0361c9f06e126c9d45c2341e6aa8. So no scope changes for this issue left.

**Comment by asr** [comment]
GitHub automatically closed the issue, so I'm reopening it.

**Comment by andreasabel** [comment]
@WolframKahl: SepElemTypes seems to pass now, please try again!
Closing...

**Comment by WolframKahl** [comment]
[ListSet0.lagda.txt](https://github.com/agda/agda/files/47916/ListSet0.lagda.txt)

Now it stumbles on the next file after SepElemTypes, on uses of SepElemTypes modules --- trimmed-down version of that file only attached (this works with previous versions).

```
 Skipping Data.SUList.SepElemTypes (/var/tmp/kahl/svn/C/RATH-Agda/trunk/Data/SUList/SepElemTypes.agdai).
/var/tmp/kahl/svn/C/RATH-Agda/trunk/Data/SUList/ListSet0.lagda:73,34-41
((Key StrictTotalOrder.≈ _x_48 Key) (_x_48 Key)) !=<
(K → K → Set (_ℓR_47 Key)) of type (Set ℓk₁)
when checking that the expression ≈K-refl has type
K → K → Set (_ℓR_47 Key)
```

**Comment by andreasabel** [comment]
Agda's last words with `-v tc.mod.apply:80`

```
Common prefix: Data.SUList.ListSet0.SepElemTypes.M1.InclEquivZip1
Copying section Data.SUList.ListSet0.SepElemTypes.M1.InclEquivZip1._ to Data.SUList.ListSet0.Zip⊆≈._
  ts           = {@3}; {@2}; {@1}; @0; {@3}; 
                   @0 .Relation.Binary.StrictTotalOrder.Carrier; 
                   Function.id {@3} {@0 .Relation.Binary.StrictTotalOrder.Carrier}; 
                   {@2}; 
                   Data.SUList.ListSet0._._≈K_ {@3} {@2} {@1} @0; 
                   λ ".e\8321" -> λ ".e\8322" -> λ "p" -> @0; 
                   {@2}; 
                   Data.SUList.ListSet0._._≈K_ {@3} {@2} {@1} @0; 
                   λ ".e\8321" -> λ ".e\8322" -> λ "p" -> @0
  tel          = ℓK ℓk₁ ℓk₂ Key ℓE Elem key ℓQ Q QK ℓR R RK
  ptel         = ℓK ℓk₁ ℓk₂ Key ℓE Elem key ℓQ Q QK ℓR R RK
  totalArgs    = 7
  parentParams = 13
  childParams  = 0
  argsToChild  = 0
  fv           = 0
```

As `totalArgs` is 7, Agda decides to drop all arguments following the first `{@2}`.  But not all of them are vanilla variables, so actual content is dropped.

@UlfNorell : Should not all dropped arguments always be certain variables, like in this case, de Bruijn indices 5,4,3,2,1,0?  I think this should be checked before dropping stuff, the relevant __IMPOSSIBLE__s need to be added. Otherwise, Agda might just silently produce garbage, no!?

The crash happens in `TC.Rules.Decl.checkSectionApplication'`

```
applying section Zip⊆.InclZipUnion
  args = ≈K-refl λ e₁ e₂ → ≈K-sym
  ptel =
  tel  = {ℓK = ℓK₁ : Level} {ℓk₁ = ℓk₃ : Level} {ℓk₂ = ℓk₄ : Level}
         (Key₁ : StrictTotalOrder ℓK₁ ℓk₃ ℓk₄) {ℓR : Level}
         (R
          : StrictTotalOrder.Carrier Key₁ →
            StrictTotalOrder.Carrier Key₁ → Set ℓR)
         (RK
          : {e₁ e₂ : StrictTotalOrder.Carrier Key₁} →
            R e₁ e₂ →
            (.Data.SUList.STOUtils._.Eq.setoid Key₁ Relation.Binary.Setoid.≈
             id e₁)
            (id e₂))
         (R-refl : {e : StrictTotalOrder.Carrier Key₁} → R e e)
         (Rf₂
          : (e₁ e₂ : StrictTotalOrder.Carrier Key₁) →
            (.Data.SUList.STOUtils._.Eq.setoid Key₁ Relation.Binary.Setoid.≈
             id e₁)
            (id e₂) →
            R e₂ e₁)
  tel' = {ℓR : Level} (R : K → K → Set ℓR)
         (RK : {e₁ e₂ : K} → R e₁ e₂ → id e₁ ≈K id e₂)
         (R-refl : {e : K} → R e e)
         (Rf₂ : (e₁ e₂ : K) → id e₁ ≈K id e₂ → R e₂ e₁)
  tel''= {ℓR : Level} (R : K → K → Set ℓR)
         (RK : {e₁ e₂ : K} → R e₁ e₂ → id e₁ ≈K id e₂)
  eta  = (R-refl : {e : K} → R e e)
         (Rf₂ : (e₁ e₂ : K) → id e₁ ≈K id e₂ → R e₂ e₁)
```

Here, `freeVarsToApply` returned apparently the wrong result; while the arguments should be checked against what is in `eta`, they are checked against what is in `tel'` (or `tel''`), crashing.

**Comment by UlfNorell** [comment]
Here's a shrunk test case. It's got something to do with datatype modules (which I did note in 2afc6a2 are treated a bit differently), because if you remove the completely unused datatype in M2 everything works.

``` agda
    module _ where

    data X : Set where
    data R (x : X) : Set where

    module SUListSepElemTypes where
      module M2 (Elem : Set) where
        data InclWith : Set where

      module M1 where
        module InclZip1 (R : X → Set) where
          open M2 X public
          module InclZipUnion (Y : Set) where

    module SepElemTypes = SUListSepElemTypes
    module Zip⊆ = SepElemTypes.M1.InclZip1 R
    module Zip⊆∪ = Zip⊆.InclZipUnion X

    -- Set !=< (X → Set) of type Set₁
    -- when checking that the expression X has type X → Set
```

**Comment by andreasabel** [comment]
This also looks wrong:

``` agda
postulate
  U V W X Y Z : Set
  u : U
  v : V
  w : W
  x : X
  y : Y
  z : Z

module Top (u : U) where
  module A (v : V) where
    module M (w : W) where
      module O (x : X) where
  module B (y : Y) where
    open A public

module ShouldFail0 where
  module C = Top.B u y
  module D = C.M.O u v w x  -- u should not be there

module ShouldFail1 where
  module C (i g n o r i n g m e : Z) = Top.B u y
  module D = C.M.O u v w x  -- u should be 10*z

module ShouldFail2 where
  module C (y : Y) = Top.B u y
  module D = C.M.O u v w x -- u should be y

module ShouldSucceedButFails where
  module C (z : Z) = Top.B u y
  module D = C.M.O z v w x -- Z !=< U of type Set 
```

**Comment by UlfNorell** [comment]
It is indeed the copying of parent modules that breaks down for datatype modules. What happens is that the parent of a datatype module (the anonymous module M2 X in this case) gets copied to the wrong thing overwriting an existing module.

**Comment by UlfNorell** [comment]
I can fix the datatype module issue, but only on master. In #836 we changed datatype/record modules so you can't use the type name as a module, which means there's no need for special copying of datatype/record modules. This change is only on master though.

@andreasabel: your examples look unrelated, maybe open a separate issue for those?

**Comment by asr** [comment]
> I can fix the datatype module issue, but only on master. In #836 we changed datatype/record modules so you can't use the type name as a module, which means there's no need for special copying of datatype/record modules. This change is only on master though.

Why #836 wasn't fixed in maint?

Anyway, I would like to release a new version after the release of GHC 7.10.3 which should soon (see https://mail.haskell.org/pipermail/ghc-devs/2015-November/010592.html).

At the moment, the _regression_ in this issue is the only blocking-issue for releasing a new version of Agda. I'll try to back-port the fix to #836.

**Comment by UlfNorell** [comment]
The reason the fix to #836 isn't on maintenance is that it changes the semantics of using/hiding, and consequently breaks people's code. I would rather we tried to find a different way of fixing this issue that doesn't rely on 836.

**Comment by asr** [comment]
> The reason the fix to #836 isn't on maintenance is that it changes the semantics of using/hiding, and consequently breaks people's code.

I see.

I know you want to release Agda 2.4.4 soon and I really agree, but I'm seeing that getting a "stable" version from the master branch is too difficult because we continue adding new features and creating new issues in this branch. Any ideas how we can address it? Maybe we can create a maint-2.4.4 branch for preparing the release of Agda 2.4.4, but I'm not sure if it is a good idea.

**Comment by asr** [comment]
BTW, since there are many and nice new features implemented in the master branch, what about to call the next release Agda 3.0.0?

**Comment by andreasabel** [comment]
We can rename 2.4.3 to 2.4.5 and release maint-2.4.2 + some stable semantics changes as 2.4.4, which would include #836 then.

**Comment by andreasabel** [comment]
@asr Agda 3.0.0 is reserved for a new language.  We are developing Agda 2, so all versions start with 2.  But we could release 2.5.0.

**Comment by asr** [comment]
> Agda 3.0.0 is reserved for a new language. We are developing Agda 2, so all versions start with 2.

Really? I think that the Agda 1 and Agda 2 distinction currently isn't necessary. Moreover, I would expect that a new language based on Agda would have a new name.

**Comment by UlfNorell** [comment]
In any case I don't think we have that many new features for 2.4.4... What were you thinking specifically?

**Comment by asr** [comment]
I proposed to bump the version after reading https://github.com/agda/agda/blob/master/doc/release-notes/2-4-4.txt.

Anyway, it is more important to define the work-flow for preparing the next release.

**Comment by asr** [comment]
> We can rename 2.4.3 to 2.4.5 and release maint-2.4.2 + some stable semantics changes as 2.4.4, which would include #836 then.

What did you mean for "We can rename 2.4.3 to 2.4.5"?

**Comment by andreasabel** [comment]
I meant that if we cannot release master now as 2.4.4, we can pick some stable feature onto the maint branch and release it as 2.4.4, and continue developing master as 2.4.5 (which will then be released as 2.4.6 or 2.5.0).

**Comment by asr** [comment]
I created #1727 for continuing the discussion.

**Comment by WolframKahl** [comment]
For the record: At 0f44002666b2c0e1f1f01907fe2f91c7bd5a5711 , I threw all my recent/active code at it and encountered no problems.

**Comment by andreasabel** [comment]
This was before the fix of #1692.  I can try to make the fix more backwards compatible (and thus, less principled, but maybe more useful).

**Comment by asr** [comment]
@WolframKahl Please add a link to your code which type-checks with Agda 2.4.2.2 but doesn't type-check with the current maintenance version.

**Comment by WolframKahl** [comment]
That link is already there --- look for “Wire” above, and note also Andreas' test case split over several files.

**Comment by UlfNorell** [comment]
This should now be fixed also on maint-2.4. I got rid of the free-variables tracking for sections completely. The same information is stored in the length of the parameter telescopes, so it's not actually needed.

(Why didn't 51248b3 auto-close the issue? Is it waiting for Travis?)

EDIT: maybe because I managed to leave that commit out of the merge...

**Comment by asr** [comment]
> Please add a link to your code which type-checks with Agda 2.4.2.2 but doesn't type-check with the current maintenance version.
> 
> > That link is already there --- look for “Wire” above

The code at http://relmics.mcmaster.ca/RATH-Agda/MonoidalCategory-Outer-Wire-2015-11-24.tar.gz doesn't type-check with Agda 2.4.2.2:

```
$ agda -i<std-lib-path> Categoric/MonoidalCategory/Outer/Wire.lagda
...
/home/asr/tmp-asr/MonoidalCategory-Outer-Wire-2015-11-24/Categoric/Category/FinColimits.lagda:146,20-25
((r
  : Iso (_A_346 C I-isInit I-isStrictInit J-isInit)
    (_B_347 C I-isInit I-isStrictInit J-isInit)) →
 CategoryProps.isIso r CategoryProps.IsIso.⁻¹ ⨾
 CategoryProps.isoMor r
 ≈ Id)
!=<
((Hom (_B_341 C I-isInit I-isStrictInit J-isInit)
  (_C_342 C I-isInit I-isStrictInit J-isInit)
  .Relation.Binary.Setoid.≈
  _i_344 C I-isInit I-isStrictInit J-isInit)
 (_j_345 C I-isInit I-isStrictInit J-isInit))
of type (Set (k ⊍ j))
when checking that the expression J→I→J has type
_i_344 C I-isInit I-isStrictInit J-isInit ≈
_j_345 C I-isInit I-isStrictInit J-isInit
```

**Comment by asr** [comment]
> This should now be fixed also on maint-2.4.

Great. Thanks!

@WolframKahl FYI, the code at  http://relmics.mcmaster.ca/RATH-Agda/MonoidalCategory-Outer-Wire-2015-11-24.tar.gz type-checked with the current version on maint.2.4.
